pkg/counter/counter.go

```diff
@@ -1,96 +1,84 @@
 package counter
 
 import (
+	"errors"
 	"fmt"
+	"math"
 	"strconv"
 	"sync/atomic"
-
-	"github.com/aquasecurity/tracee/pkg/errfmt"
 )
 
 type Counter struct {
-	count uint64
+	value uint64
 }
 
-func NewCounter(v uint64) Counter {
-	return Counter{
-		count: v,
-	}
+// NewCounter creates a new counter with the given initial value.
+func NewCounter(initialValue uint64) Counter {
+	return Counter{value: initialValue}
 }
 
-func (c *Counter) Increment(amount ...uint64) error {
-	if len(amount) == 0 {
-		return c.incAtomic(1)
-	}
-
-	sum, err := sumUint64(amount...)
-	if err != nil {
-		return errfmt.WrapError(err)
-	}
-
-	return c.incAtomic(sum)
+// Increase increases the counter by the given value (thread-safe).
+func (c *Counter) Increase(values ...uint64) error {
+	_, err := c.IncreaseAndRead(values...)
+	return err
 }
 
-func (c *Counter) Decrement(amount ...uint64) error {
-	if len(amount) == 0 {
-		return c.decAtomic(1)
+// IncreaseAndRead increases the counter by the given value (thread-safe) and returns the new value.
+func (c *Counter) IncreaseAndRead(values ...uint64) (uint64, error) {
+	if len(values) == 0 {
+		values = append(values, 1)
 	}
 
-	sum, err := sumUint64(amount...)
-	if err != nil {
-		return errfmt.WrapError(err)
-	}
+	var n uint64
 
-	return c.decAtomic(sum)
-}
-
-func (c *Counter) Read() uint64 {
-	return atomic.LoadUint64(&c.count)
-}
+	for _, value := range values {
+		n = atomic.AddUint64(&c.value, value)
+		if n < value {
+			return n, errors.New("counter overflow")
+		}
+	}
```

Comment:
It is better to first sum the values (as was done previously) and then increment than incrementing for every value since:
1. sum is a local variable, thus not exposed to concurrency issues
2. keeping things simple and not using atomic for every value
3. theoretically there is a race condition in the overflow check when two callers have multiple values - if the first caller overflows and we context switch before checking `n < value` then the second caller increments and context switch back to the first caller, where the overflow condition is not necessarily true anymore

Reply:
You can't "not use atomics" and still be race free, because of HW cache coherence. You are right about the possible race (and it was known when I write). Its better to have 2 errors than 2 increase/decreases.

Reply:
There is no usecase (1) in Tracee, so we are safe to use atomics only and check for overflow after atomic operation happened. Will put a PR together.

Reply:
https://github.com/aquasecurity/tracee/pull/3276 I believe it address all of our concerns and is race-free without mutexes. Suggestions from everybody =).